mycpp/pass_state.py

```diff
@@ -485,3 +541,40 @@ def DumpControlFlowGraphs(cfgs: dict[str, ControlFlowGraph],
 
     for f in fact_files.values():
         f.close()
+
+
+def ComputeMinimalStackRoots(cfgs: dict[str, ControlFlowGraph],
+                      facts_dir: str = '_tmp/mycpp-facts',
+                      souffle_output_dir: str = '_tmp') -> StackRoots:
+    """
+    Run the the souffle stack roots solver and translate its output in a format
+    that can be queried by cppgen_pass.
+    """
+    DumpControlFlowGraphs(cfgs, facts_dir=facts_dir)
+    subprocess.check_call([
+        '_bin/datalog/dataflow',
+        '-F',
+        facts_dir,
+        '-D',
+        souffle_output_dir,
+    ])
+
+    tuples: set[tuple[SymbolPath, SymbolPath]] = set({})
+    with open('{}/stack_root_vars.tsv'.format(souffle_output_dir),
+              'r') as roots_f:
+        pat = re.compile(r'\$(.*)\((.*), (.*)\)')
+        for line in roots_f:
+            function, ref = line.split('\t')
+            m = pat.match(ref)
+            assert m.group(1) in ('LocalVariable', 'ObjectMember')
+            if m.group(1) == 'LocalVariable':
+                _, ref_func, var_name = m.groups()
+                assert ref_func == function
+                tuples.add((split_py_name(function), (var_name, )))
+
+            if m.group(1) == 'ObjectMember':
```

Comment:
OK I guess we need  ObjectMember only for context managers?

Reply:
Yup! Maybe we'll use them for something else in the future, but for now that's the only case we care about.